docker_push_latest_if_changed.py

```diff
@@ -1,5 +1,182 @@
+    target = _get_sanitized_target(source, target)
+
+    logging.info('Pushing source image')
+    _push_image(source, docker_client, is_dry_run)
+    try:
+        logging.warning('Pulling target image')
+        docker_client.pull(target)
+    except docker.errors.NotFound:
+        logging.warning((
+            'Target image {} was not found in the registry. '
+            'Going to attempt to tag and push the target image anyway.'
+        ).format(target))
+        _tag_image(source, target, docker_client, is_dry_run)
+        _push_image(target, docker_client, is_dry_run)
+    else:
+        if _is_image_changed(source, target, docker_client):
+            logging.warning((
+                'Source image was found to be different from the current '
+                'version of the target image. Tagging and pushing new target '
+                'image.'
+            ))
+            _tag_image(source, target, docker_client, is_dry_run)
+            _push_image(target, docker_client, is_dry_run)
+        else:
+            logging.warning((
+                'Source image was found to be the same as the current '
+                'version of the target image. Not pushing the target image.'
+            ))
+
+
+def _validate_source(source, docker_client):
+    logging.debug('Docker inspecting source to verify it exists')
+    source_inspect = docker_client.inspect_image(source)
+    if source not in source_inspect['RepoTags']:
+        # Hack to check if source is just a repo without a tag.
+        # An inspect_image query with just a repo will search for
+        # latest, but the result in RepoTags will only have
+        # repo:latest and not just repo
+        raise ValueError((
+            'The source image {source} does not have a tag! '
+            'You must include a tag in the source parameter.'
+        ).format(source=source))
+
+
+def _get_sanitized_target(source, target):
+    if not target:
+        repository = source.rsplit(':', 1)[0]
+        target = '{repository}:latest'.format(repository=repository)
+        logging.info((
+            'Target was not given so falling back to ":latest". '
+            'Target is now {}'
+        ).format(target))
+    if source == target:
+        raise ValueError((
+            'The source ({source}) and target ({target}) repo:tags are both '
+            'the same! Source and target cannot be the same.'
+        ).format(source=source, target=target))
+    return target
+
+
+def _tag_image(source, target, docker_client, is_dry_run):
+    logging.warning('Tagging target image {}'.format(target))
+    if is_dry_run:
+        logging.warning(
+            'WARNING: Image was not actually tagged since this is a dry run'
+        )
+    else:
+        docker_client.tag(source, target)
+
+
+def _push_image(image, docker_client, is_dry_run):
+    logging.warning('Pushing target image {} ...'.format(image))
+    if is_dry_run:
+        logging.warning(
```

Comment:
I don't think this should be a warning case, if the user asked for dry run of course it's not going to push

Reply:
Yeah, a warning really is spooky and shouldn't be used here.
This should be resolved after switching to verbose level prints rather than depending on logging levels.